x-pack/plugins/security_solution/server/endpoint/services/actions/clients/crowdstrike/crowdstrike_actions_client.ts

```diff
@@ -0,0 +1,318 @@
+};
+
+export class CrowdstrikeActionsClient extends ResponseActionsClientImpl {
+  protected readonly agentType: ResponseActionAgentType = 'crowdstrike';
+  private readonly connectorActionsClient: NormalizedExternalConnectorClient;
+
+  constructor({ connectorActions, ...options }: CrowdstrikeActionsClientOptions) {
+    super(options);
+    this.connectorActionsClient = connectorActions;
+    connectorActions.setup(CROWDSTRIKE_CONNECTOR_ID);
+  }
+
+  protected async writeActionRequestToEndpointIndex<
+    TParameters extends EndpointActionDataParameterTypes = EndpointActionDataParameterTypes,
+    TOutputContent extends EndpointActionResponseDataOutput = EndpointActionResponseDataOutput,
+    TMeta extends {} = {}
+  >(
+    actionRequest: ResponseActionsClientWriteActionRequestToEndpointIndexOptions<
+      TParameters,
+      TOutputContent,
+      TMeta
+    >
+  ): Promise<
+    LogsEndpointAction<TParameters, TOutputContent, TMeta & CrowdstrikeActionRequestCommonMeta>
+  > {
+    const agentId = actionRequest.endpoint_ids[0];
+    const eventDetails = await super.getEventDetailsById<{
+      crowdstrike: { event: { HostName: string } };
+    }>({
+      index: ['logs-crowdstrike.fdr*', 'logs-crowdstrike.falcon*'],
+      term: { 'crowdstrike.event.DeviceId': agentId },
+      fields: [{ field: 'crowdstrike.event.HostName' }],
+    });
+
+    const hostname = eventDetails.crowdstrike.event.HostName;
+    return super.writeActionRequestToEndpointIndex({
+      ...actionRequest,
+      hosts: {
+        [agentId]: { name: hostname },
+      },
+      meta: {
+        hostName: hostname,
+        agentId,
+        ...(actionRequest.meta ?? {}),
+      } as TMeta & CrowdstrikeActionRequestCommonMeta,
+    });
+  }
+
+  /**
+   * Sends actions to Crowdstrike directly (via Connector)
+   * @private
+   */
+  private async sendAction(
+    actionType: SUB_ACTION,
+    actionParams: object
+  ): Promise<ActionTypeExecutorResult<unknown>> {
+    const executeOptions: Parameters<typeof this.connectorActionsClient.execute>[0] = {
+      params: {
+        subAction: actionType,
+        subActionParams: actionParams,
+      },
+    };
+
+    this.log.debug(
+      `calling connector actions 'execute()' for Crowdstrike with:\n${stringify(executeOptions)}`
+    );
+
+    const actionSendResponse = await this.connectorActionsClient.execute(executeOptions);
+
+    if (actionSendResponse.status === 'error') {
+      this.log.error(stringify(actionSendResponse));
+
+      throw new ResponseActionsClientError(
+        `Attempt to send [${actionType}] to Crowdstrike failed: ${
+          actionSendResponse.serviceMessage || actionSendResponse.message
+        }`,
+        500,
+        actionSendResponse
+      );
+    }
+
+    this.log.debug(`Response:\n${stringify(actionSendResponse)}`);
+
+    return actionSendResponse;
+  }
+
+  // TODO TC: uncomment when working on agent status support
+  // private async getAgentDetails(
```

Comment:
So you're suggesting that I do not expose it through a protected method in base client but just a private method in Crowdstrike client? 

That's fine, just thought it could be reused in S1 too. I'll refactor, thanks!

Reply:
On second thoughts I think it would be a different method. The previous one just gets the hostname from ES. This here would be used to get the agentStatus. 
I'll prepare a second method then 👍

Reply:
the `getAgentDetails()` method of SentinelOne client sends an API call to S1 to get the full details of the Agent/host. It could (I guess) in the future be changed to instead query the internal index, but the purpose of the method remains the same: retrieve the details of the agent from S1